<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html xmlns:cis="http://zlab.bu.edu/schema/cisml" xmlns:fimo="http://noble.gs.washington.edu/schema/cisml" xmlns:mem="http://noble.gs.washington.edu/meme">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="UTF-8">
<title>FIMO Results</title>
<style type="text/css">
        td.left {text-align: left;}
        td.right {text-align: right; padding-right: 1cm;}
      </style>
</head>
<body bgcolor="#D5F0FF">
<a name="top_buttons"></a><hr>
<table summary="buttons" align="left" cellspacing="0"><tr>
<td bgcolor="#00FFFF"><a href="#database_and_motifs"><b>Database and Motifs</b></a></td>
<td bgcolor="#DDFFDD"><a href="#sec_i"><b>High-scoring Motif Occurences</b></a></td>
<td bgcolor="#DDDDFF"><a href="#debugging_information"><b>Debugging Information</b></a></td>
</tr></table>
<br><br><hr>
<center><big><b>FIMO - Motif search tool</b></big></center>
<hr>
<p>
      FIMO version 4.10.0, 
      (Release date: Tue May 06 16:46:04 2014 +1000)
    </p>
<p>
      For further information on how to interpret these results
      or to get a copy of the FIMO software please access
      <a href="http://meme.nbcr.net">http://meme.nbcr.net</a></p>
<p>If you use FIMO in your research, please cite the following paper:<br>
      Charles E. Grant, Timothy L. Bailey, and William Stafford Noble,
      "FIMO: Scanning for occurrences of a given motif",
      <i>Bioinformatics</i>, <b>27</b>(7):1017-1018, 2011.
      <a href="http://bioinformatics.oxfordjournals.org/content/27/7/1017">[full text]</a></p>
<hr>
<center><big><b><a name="database_and_motifs">DATABASE AND MOTIFS</a></b></big></center>
<hr>
<div style="padding-left: 0.75in; line-height: 1em; font-family: monospace;">
<p>
        DATABASE 
        memechip_example_output_files/sample-dna-Klf1.fa<br>
        Database contains 
        904
        sequences,
        452000
        residues
      </p>
<p>
        MOTIFS 
        memechip_example_output_files/dreme_out/dreme.xml 
        (nucleotide)
        <table>
<thead><tr>
<th style="border-bottom: 1px dashed;">MOTIF</th>
<th style="border-bottom: 1px dashed; padding-left: 1em;">WIDTH</th>
<th style="border-bottom: 1px dashed; padding-left: 1em;" align="left">
              BEST POSSIBLE MATCH
            </th>
</tr></thead>
<tbody>
<tr>
<td align="right">GGGYGK</td>
<td align="right" style="padding-left: 1em;">6</td>
<td align="left" style="padding-left: 1em;">GGGTGT</td>
</tr>
<tr>
<td align="right">TTATCW</td>
<td align="right" style="padding-left: 1em;">6</td>
<td align="left" style="padding-left: 1em;">TTATCT</td>
</tr>
<tr>
<td align="right">AGAWA</td>
<td align="right" style="padding-left: 1em;">5</td>
<td align="left" style="padding-left: 1em;">AGAAA</td>
</tr>
</tbody>
</table></p>
<p>
        Random model letter frequencies 
        (from memechip_example_output_files/background):
        <br>A 0.256 C 0.244 G 0.244 T 0.256 </p>
</div>
<hr>
<center><big><b><a name="sec_i">SECTION I: HIGH-SCORING MOTIF OCCURENCES</a></b></big></center>
<hr>
<ul>
<li>
    There were
    0
    motif occurences with a
    
        p-value less than
    0.0001.
    </li>
<li> The p-value of a motif occurrence is defined as the
    probability of a random sequence of the same length as the motif
    matching that position of the sequence with as good or better a score.
    </li>
<li> The score for the match of a position in a sequence to a motif
    is computed by summing the appropriate entries from each column of
    the position-dependent scoring matrix that represents the motif.
    </li>
<li>The table is sorted by increasing p-value.</li>
</ul>
<table border="1">
<thead><tr>
<th>Motif</th>
<th>Sequence Name</th>
<th>Strand</th>
<th>Start</th>
<th>End</th>
<th>p-value</th>
<th>Matched Sequence</th>
</tr></thead>
<tbody></tbody>
</table>
<hr>
<center><big><b><a name="debugging_information">DEBUGGING INFORMATION</a></b></big></center>
<hr>
<p>
    Command line:
    </p>
<pre>/Users/cegrant/meme_4.10.0/bin/fimo --parse-genomic-coord --verbosity 1 --oc memechip_example_output_files/fimo_out_1 --bgfile memechip_example_output_files/background --motif GGGYGK memechip_example_output_files/dreme_out/dreme.xml memechip_example_output_files/sample-dna-Klf1.fa</pre>
<p>
    Settings:
    </p>
<pre><table>
<tr>
<td style="padding-right: 2em">output directory = memechip_example_output_files/fimo_out_1</td>
<td style="padding-left: 5em; padding-right: 2em">MEME file name = memechip_example_output_files/dreme_out/dreme.xml</td>
<td style="padding-left: 5em; padding-right: 2em">sequence file name = memechip_example_output_files/sample-dna-Klf1.fa</td>
</tr>
<tr>
<td style="padding-right: 2em">background file name = memechip_example_output_files/background</td>
<td style="padding-left: 5em; padding-right: 2em">allow clobber = true</td>
<td style="padding-left: 5em; padding-right: 2em">compute q-values = true</td>
</tr>
<tr>
<td style="padding-right: 2em">parse genomic coord. = true</td>
<td style="padding-left: 5em; padding-right: 2em">text only = false</td>
<td style="padding-left: 5em; padding-right: 2em">scan both strands = true</td>
</tr>
<tr>
<td style="padding-right: 2em">max sequence length = 250000000</td>
<td style="padding-left: 5em; padding-right: 2em">output threshold = 0.0001</td>
<td style="padding-left: 5em; padding-right: 2em">threshold type = p-value</td>
</tr>
<tr>
<td style="padding-right: 2em">max stored scores = 100000</td>
<td style="padding-left: 5em; padding-right: 2em">pseudocount = 0.1</td>
<td style="padding-left: 5em; padding-right: 2em">verbosity = 1</td>
</tr>
<tr>
<td style="padding-right: 2em">selected motif = GGGYGK</td>
<td style="padding-left: 5em; padding-right: 2em"></td>
<td align="right"></td>
</tr>
</table></pre>
<p>
      This information can be useful in the event you wish to report a
      problem with the FIMO software.
    </p>
<hr>
<span style="background-color: #DDDDFF"><a href="#top_buttons"><b>Go to top</b></a></span>
</body>
</html>
