<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>AMA - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
    <script src="js/shared-doc.js"></script>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>

    <div class="prog_logo">
      <!--<img src="images/ama_icon.png" alt="AMA Logo">-->
      <h1>AMA</h1>
      <h2>Average Motif Affinity</h2>
    </div>

    <div class="usage">
      <h3>Usage:</h3> 
      <code>ama [options] &lt;motif file&gt; &lt;sequence file&gt;
        [&lt;background file&gt;]</code>
    </div>
    <h3>Description</h3>
    <div class="indent">
      <p>The name AMA stands for "Average Motif Affinity". The program scores
      a set of DNA sequences given a DNA-binding motif, treating each position
      in the sequence as a possible binding event. The score is calculated by
      averaging the likelihood ratio scores for all feasible binding events to
      the given sequence and to its reverse strand. The binding strength at
      each potential site is defined as the likelihood ratio of the site under
      the motif versus under a zero-order background model provided by the
      user.</p>
      <p>By default, AMA reports the average motif affinity score. It can also
      report <i>p</i>-values, which are estimated analytically using the given
      zero-order background model or using the GC-content of each
      sequence.</p>
      <p>AMA can also compute the sequence-dependent likelihood ratio score
      used by Clover. The denominator of this score depends on the sequence
      being scored, and is the likelihood of the site under a Markov model
      derived from the sequence itself.  Unlike Clover, AMA also allows
      higher-order sequence-derived Markov models (see
      <span class="popt">--sdbg</span> option below).</p>
      <p>If the input file contains more than one motif, the motifs will be
      processed consecutively.</p>
      <p>Full details are given in the supplement to the GOMO paper:</p>
      <span class="citation">
      Fabian A. Buske, Mikael Bod&eacute;n, Denis C. Bauer and Timothy L. Bailey, 
      "<a href="http://bioinformatics.oxfordjournals.org/content/26/7/860.full"
      >Assigning roles to DNA regulatory motifs using comparative genomics</a>",
      <em>Bioinformatics</em>, 26(7):860-866, 2010.
      </span>
    </div>
    <h3>Inputs</h3>
    <div class="indent">
      <h4>Motif File</h4>
      <p>A file containing a list of motifs, in
      <a href="meme-format.html">MEME format</a>.</p>
      <h4>Sequence File</h4>
      <p>A file containing a collection of sequences in
      <a href="fasta-format.html">FASTA</a> format.</p>
      <h4>Background File</h4>
      <p>A file containing 0-order Markov Model in
      <a href="bfile-format.html">background model format</a> such as produced
      by <a href="fasta-get-markov.html"><code>fasta-get-markov</code></a>.</p>
      <p>This is a required option unless <span class="popt">--sdbg</span> is
      specified.</p>
    </div>
    <h3>Outputs</h3>
    <div class="indent">
      <p>AMA writes to standard out, unless you specify one of
      <span class="popt">--o</span> or <span class="popt">--oc</span> in which
      case the <span class="popt">o-format</span> option (if given) is ignored
      and separate files containing each output format are written to the named
      directory. The available output formats are
      <a href="http://www.sanger.ac.uk/Software/formats/GFF">gff</a> and
      <a href="http://zlab.bu.edu/~phaverty/SupplementalData/CisML">CisML</a>.
      </p>
      <p>gff output has the format: 
      <pre> &lt;sequence_name&gt; ama sequence 1 &lt;sequence_length&gt; &lt;sequence_score&gt; &lt;sequence_p-value&gt; . . .</pre>
      </p>
    </div>
    <h3>Options</h3>
    <div class="indent">
      <table class="options">
        <thead>
          <tr>
            <th>Option</th><th>Parameter</th><th>Description</th><th>Default Behaviour</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <th colspan="4">General Options</th>
          </tr>
          <tr>
            <td class="popt">--sdbg</td><td><span class="pdat">n</span></td>
            <td>Use a sequence-dependent Markov model of order
            <span class="pdat">n</span> when computing likelihood ratios. A
            different sequence-dependent Markov model is computed for each
            sequence in the input and used to compute the likelihood ratio of
            all sites in that sequence. This option overrides
            <span class="popt">--pvalues</span>,
            <span class="popt">--gcbins</span>, and
            <span class="popt">--rma</span>.</td>
            <td>The background file is required and is used to compute the
            likelihood ratio for all sites in all sequences.</td>
          </tr>
          <tr>
            <td class="popt">--motif</td><td><span class="pdat">id</span></td>
            <td>Use only the motif identified by
            <span class="pdat">id</span>. This option may be repeated.</td>
            <td>All motifs are used.</td>
          </tr>
          <tr>
            <td class="popt">--motif-pseudo</td><td><span class="pdat">float</span></td>
            <td>A pseudocount to be added to each count in the motif matrix,
            after first multiplying by the corresponding background
            frequency.</td>
            <td>A pseudocount of 0.1 is applied.</td>
          </tr>
          <tr>
            <td class="popt">--norc</td><td>&nbsp;</td>
            <td>Do not score the reverse complement DNA strand.</td>
            <td>Both strands are scored.</td>
          </tr>
          <tr>
            <td class="popt">--scoring</td><td><span class="popt">avg-odds</span>|<span class="popt">max-odds</span></td>
            <td>Indicates whether the average or the maximum likelihood ratio
            (odds) score should be calculated. If max-odds is chosen, no 
            <i>p</i>-value will be printed.</td>
            <td>Average score will be calculated.</td>
          </tr>
          <tr>
            <td class="popt">--rma</td><td>&nbsp;</td>
            <td>Scale the motif affinity score by the maximum achievable score
            for each motif. This is termed the <b>R</b>elative <b>M</b>otif
            <b>A</b>ffinity score. This allows for direct comparison between
            different motifs.</td>
            <td>Affinity scores are not scaled.</td>
          </tr>
          <tr>
            <td class="popt">--pvalues</td><td>&nbsp;</td>
            <td>Print the <i>p</i>-value of the average odds score in the
            output file. The <i>p</i>-score for a score is normally computed
            (but see <span class="popt">--gcbins</span>) assuming the
            sequences were each generated by the 0-order Markov model
            specified by the background file frequencies. This option is
            ignored if max-odds scoring is used.</td>
            <td>No <i>p</i>-value will be printed.</td>
          </tr>
          <tr>
            <td class="popt">--gcbins</td><td><span class="pdat">bins</span></td>
            <td>Compensate <i>p</i>-values for the GC content of each sequence
            independently. This is done by computing the score distributions
            for a range of GC values.  Using 41 bins (recommended) computes
            distributions at intervals of 2.5% GC content. The computation
            assumes that the ratios of G to C and A to T are both equal to 1.
            This assumption will fail if a sequence contains far more of a
            letter than its complement. This option sets the
            <span class="popt">--pvalues</span> option. This option is ignored
            if max-odds scoring is used.</td>
            <td>Uncompensated <i>p</i>-values are printed.</td>
          </tr>
          <tr>
            <td class="popt">--cs</td><td>&nbsp;</td>
            <td>Enables combining of sequences with the same identifier by
            taking the average score and the Sidak corrected <i>p</i>-value:
            1&minus;(1&minus;&alpha;)^1/n. Different sequences with the same identifier are used
            in GOMO databases if one gene in the reference species has more
            than one homologous gene in the related species (one-to-many
            relationship).</td>
            <td>Sequences are processed independently of each other.</td>
          </tr>
          <tr>
            <td class="popt">--o-format</td><td><span class="popt">gff</span>|<span class="popt">cisml</span></td>
            <td>Set the output file format.</td>
            <td>CISML output format is used.</td>
          </tr>
          <tr>
            <td class="popt">--o</td><td><span class="pdat">dir</span></td>
            <td>Create a folder called <span class="pdat">dir</span> and write
            output files in it. This option is not compatible with -oc as only one
            output folder is allowed.</td>
            <td>The program writes to standard out.</td>
          </tr>
          <tr>
            <td class="popt">--oc</td><td><span class="pdat">dir</span></td>
            <td>Create a folder called <span class="pdat">dir</span> but if it
            already exists allow overwriting the contents. This option is not
            compatible with -o as only one output folder is allowed.</td>
            <td>The program writes to standard out.</td>
          </tr>
          <script>
            wropt("all-o", {"ddash": true});
            wropt("all-oc", {"ddash": true});
            wropt("all-verbosity", {"ddash": true});
          </script>
          <tr>
            <td class="popt">--max-seq-length</td><td><span class="pdat">max</span></td>
            <td>Set the maximum length allowed for input sequences to
            <span class="pdat">max</span>.</td>
            <td>The maximum allowed input sequence length is 250000000.</td>
          </tr>
          <tr>
            <td class="popt">--last</td><td><span class="pdat">n</span></td>
            <td>Use only scores of (up to) last <span class="pdat">n</span>
            sequence positions to compute AMA. If the sequence is shorter than
            this value the entire sequence is scored. If the motif is longer
            than this value it will not be scored.</td>
            <td>The full sequence is scored.</td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>
</html>

