<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>Beadstring - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
    <script src="js/shared-doc.js"></script>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>
    <h2>Beadstring (unsupported)</h2>
    <div class="usage">
      <h3>Usage:</h3>
      <code>beadstring [options] &lt;motifs&gt; &lt;database&gt;</code>
    </div>
    <h3>Description</h3>
    <div class="indent">
      <p>Beadstring builds a linear hidden Markov model (HMM) from the motifs
      and motif occurences listed in the motif file, and uses that HMM to
      search a sequence database for a particular ordered series of motifs. A
      description of the algorithm is found in:</p>

      <span class="citation">
        <a href="http://bioinformatics.oxfordjournals.org/cgi/content/abstract/13/4/397">
          Grundy, Bailey, Elkan and Baker. "Meta-MEME: Motif-based Hidden Markov Models of Protein Families".
          <i>Computer Applications in the Biosciences.</i> 13(4):397-406, 1997.
        </a>
      </span>

      <p>By default, the order and spacing of motifs in the model is determined 
      from the "Summary of Motifs" section of the MEME input file. Beadstring
      searches the summary for the sequence that contains the maximal number of
      distinct motif occurrences. If there is a tie, then beadstring selects the
      sequence with the smallest combined <i>p</i>-value. Beadstring then
      eliminates all but the most significant occurrence of each motif and uses
      the resulting order and spacing of motif occurrences to initialize the
      HMM.  This procedure can be overridden by selecting the
      <span class="popt">--motif</span>, 
      <span class="popt">--motif-e-thresh</span>,
      <span class="popt">--motif-p-thresh</span> or 
      <span class="popt">--order</span> options.</p>
      
      <p>The command line option <span class="popt">--p-score</span> activates
      an alternative scoring mode, called "<i>p</i>-value scoring." This scoring
      method is described in:</p>

      <span class="citation">
        <a href="http://bioinformatics.oxfordjournals.org/cgi/reprint/19/suppl_2/ii16">
          Bailey and Noble. "Searching for statistically significant regulatory modules."
          <i>Bioinformatics</i> 19(Suppl 2):ii16-ii25, 2003.
        </a>
      </span>
    </div>
    <h3>Input</h3>
    <div class="indent">
      <h4>Motifs</h4>
      <p>A file of motifs from MEME (DREME does not have the 
      "Summary of Motifs" section).</p>
      <h4>Database</h4>
      <p>A database of sequences in <a href="fasta-format.html">FASTA format</a>.</p>
    </div>
    <h3>Output</h3>
    <div class="indent">
      <p>Beadstring will create a directory, named <code>beadstring_out</code>
      by default. Any existing output files in the directory will be
      overwritten. The directory will contain:</p>
      <ul>
        <li>An XML file named <code>beadstring.xml</code> using the 
        <a href="http://zlab.bu.edu/~phaverty/SupplementalData/CisML">CisML</a>
        schema.</li>
        <li>An XML file named <code>model.xml</code> using the MEME_HMM
        schema.</li>
        <li>An HTML file named <code>beadstring.html</code></li>
        <li>A plain text file named <code>beadstring.text</code></li>
      </ul>
      <p>The default output directory can be overridden using the
      <style class="popt">--o</style> or <style class="popt">--oc</style>
      options which are described below.</p>
    </div>
    <h3>Options</h3>
    <div class="indent">
      <table class="options">
        <thead>
          <tr>
            <th>Option</th>
            <th>Parameter</th>
            <th>Description</th>
            <th>Default Behaviour</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <th colspan="4">Input/Output</th>
          </tr>
          <tr>
            <td class="popt">--bgfile</td><td><span class="pdat">bfile</span></td>
            <td>Read background frequencies from <span class="pdat">bfile</span>.
            The file should be in 
            <a href="bfile-format.html">MEME background file format</a>. 
            The default is to use frequencies embedded in the application from
            the non-redundant database. If the argument is the keyword
            <code>motif-file</code>, then the frequencies will be taken from
            the motif file.</td>
            <td>Use NR frequencies.</td>
          </tr>
          <tr>
            <td class="popt">--e-thresh</td><td><span class="pdat">ev</span></td>
            <td>Only print results with E-values less than
            <span class="pdat">ev</span>.</td>
            <td>Print results with E-values less than 0.01.</td>
          </tr>
          <tr>
            <td class="popt">--max-seqs</td><td><span class="pdat">max</span></td>
            <td>Print results for no more than <span class="pdat">max</span>
            sequences.</td>
            <td>All matches are reported, up to the specified E-value
            threshold (see <span class="popt">--e-thresh</span>).
          </tr>
          <tr>
            <td class="popt">--model-file</td><td><span class="pdat">model file</span></td>
            <td>Creation of the HMM will be skipped, and the HMM will be read
            from the file instead.</td>
            <td>The HMM will be created.</td>
          </tr>
          <tr>
            <td class="popt">--no-search</td><td>&nbsp;</td>
            <td>This option turns off the search phase of beadstring.
            The HMM will be stored if the <span class="popt">--model</span>
            option is specified.</td>
            <td>The search phase runs as normal.</td>
          </tr>
          <script>
            wropt("all-o", {"ddash": true, "dir": "beadstring_out"});
            wropt("all-oc", {"ddash": true, "dir": "beadstring_out"});
          </script>
          <tr>
            <td class="popt">--progress</td><td><span class="pdat">value</span></td>
            <td>Print to standard error a progress message approximately every
            <span class="pdat">value</span> seconds.</td>
            <td>No progress message.</td>
          </tr>
          <tr>
            <td class="popt">--score-file</td><td><span class="pdat">score file</span></td>
            <td>Cause a score file (in BLAST format) to be read and used
            instead of the built-in PAM (for proteins) or
            transition/transversion (for DNA) score file. Several score files
            are provided (including BLOSUM62) in the directory
            <code>doc</code>. Other, user-provided score files may be
            specified as well, as long as they are in the proper format.</td>
            <td>Uses the built-in score file.</td>
          </tr>
          <script>wropt("all-verbosity");</script>
        </tbody>
        <tbody>
          <tr>
            <th colspan="4">Motif Selection</th>
          </tr>
          <tr>
            <td class="popt">--motif</td><td><span class="pdat">id</span></td>
            <td>Use only the motif identified by <span class="pdat">id</span>.
            This option may be repeated.</td>
            <td>Use all motifs that pass the other motif selection options.</td>
          </tr>
          <tr>
            <td class="popt">--motif-e-thresh</td><td><span class="pdat">ev</span></td>
            <td>Use only motifs with E-values less than 
            <span class="pdat">ev</span>.</td>
            <td>Use all motifs that pass the other motif selection options.</td>
          </tr>
          <tr>
            <td class="popt">--motif-p-thresh</td><td><span class="pdat">pv</span></td>
            <td>Use only motifs with <i>p</i>-values less than
            <span class="pdat">pv</span>.</td>
            <td>Use all motifs that pass the other motif selection options.</td>
          </tr>
          <tr>
            <td class="popt">--order</td><td><span class="pdat">string</span></td>
            <td>The given string specifies the order and spacing of the motifs
            within the model, and has the format "l=n=l=n=...=l=n=l", where
            "l" is the length of a region between motifs, and "n" is a motif
            index. Thus, for example, the string "34=3=17=2=5" specifies a
            two-motif linear model, with motifs 3 and 2 separated by 17
            letters and flanked by 34 letters and 5 letters on the left and
            right. If the motif file contains motif occurrences on both
            strands, then the motif IDs in the order string should be
            preceded by "+" or "-" indicating the strandedness of the
            motif.</td>
            <td>The order and spacing is determined from the motif file.</td>
          </tr>
        </tbody>
        <tbody>
          <tr>
            <th colspan="4">Building The Model</th>
          </tr>
          <tr>
            <td class="popt">--fim</td><td>&nbsp;</td>
            <td>Gaps between motifs are not penalized. Spacer states between
            motifs are represented as free-insertion modules (FIM). A FIM is
            an insert state with 1.0 probability of self-transition and 1.0
            probability of exit transition. Thus, traversing such a state has
            zero transition cost. Specifying this option causes all spacers to
            be represented using FIMs.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">--gap-extend</td><td><span class="pdat">cost</span></td>
            <td>This switch causes <b>all</b> spacer self-loop log-odds scores
            to be set to <span class="pdat">cost</span>. In addition, it
            causes all other transitions out of a spacer to be set to zero.
            Together with the <span class="popt">--gap-open</span> switch,
            this allows you to specify an affine gap penalty function,
            overriding the gap penalty implicit in the model (self-loop
            transition probabilities of gap states).</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">--gap-open</td><td><span class="pdat">cost</span></td>
            <td>This switch causes <b>all</b> transitions into a spacer state
            to be assigned a log-odds score equal to
            <span class="pdat">cost</span>. Together with the
            <span class="popt">--gap-extend</span> switch, this allows you to
            specify an affine gap penalty function, overriding the gap penalty
            implicit in the model (transition probabilities into and out of
            gap states).</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">--motif-pseudo</td><td>num</td>
            <td>A pseudocount to be added to each count in the motif matrix,
            after first multiplying by the corresponding background
            frequency.</td>
            <td>No pseudocount is added to motif matrix probabilities.</td>
          </tr>
          <tr>
            <td class="popt">--nspacer</td><td><span class="pdat">value</span></td>
            <td>By default each spacer is modeled using a single insert state.
            The distribution of spacer lengths produced by a single insert
            state is exponential in form. A more reasonable distribution would
            be a bell-shaped curve such as a Gaussian. Modeling the length
            distribution explicitly is computationally expensive; however, a
            Gaussian distribution can be approximated using multiple insert
            states to represent a single spacer region. The
            <span class="popt">--nspacer</span> option specifies the number of
            insert states used to represent each spacer.</td>
            <td>A single insert state is used.</td>
          </tr>
          <tr>
            <td class="popt">--spacer-pseudo</td><td><span class="pdat">value</span></td>
            <td>Specify the value of the pseudocount used in converting
            transition counts to spacer self-loop probabilities.</td>
            <td>No pseudocount is added to self-loop probabilities.</td>
          </tr>
          <tr>
            <td class="popt">--trans-pseudo</td><td><span class="pdat">value</span></td>
            <td>Specify the value of the pseudocount used in converting
            transition counts to transition probabilities.</td>
            <td>A pseudocount of 0.1 is added to transition probabilities.</td>
          </tr>
          <tr>
            <td class="popt">--zselo</td><td>&nbsp;</td>
            <td>Spacer emission log-odds scores to be set to zero. This
            prevents regions of unusual base/residue composition matching
            spacers well when the spacer emission frequencies are different
            than the background frequencies. It is particularly useful with
            DNA models.</td>
            <td></td>
          </tr>
        </tbody>
        <tbody>
          <tr>
            <th colspan="4">Scoring</th>
          </tr>
          <tr>
            <td class="popt">--allow-weak-motifs</td><td>&nbsp;</td>
            <td>In p-value score mode, weak motifs are defined as ones where
            the best possible hit has a p-value greater than the
            <i>p</i>-value threshold. Such motifs cannot contribute to a match
            in p-value score mode. By default, the program rejects any search
            results containing weak motifs, unless the
            <span class="popt">--allow-weak-motifs</span> switch is given. In
            that case, the search will proceed, but the weak motifs will never
            appear in any matches. <b>Note</b>:This switch only applies to
            <i>p</i>-value score mode.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">--global</td><td>&nbsp;</td>
            <td>Scores are computed for the match between the entire sequence
            and the model.</td>
            <td>Use the maximal local score.</td>
          </tr>
          <tr>
            <td class="popt">--pam</td><td><span class="pdat">distance</span></td>
            <td>By default, target probabilities are derived from the
            distance-250 PAM matrix for proteins, and from a
            <span class="pdat">distance</span>-1 transition/transversion
            matrix for DNA. With the <span class="popt">-pam</span> switch,
            you can specify a different <b>integer</b> distance from 1 to 500.
            (This can be overridden with the
            <span class="popt">--score-file</span> switch below). The
            <span class="pdat">distance</span>-1 transition/transversion joint
            probability matrix for DNA is given below:
            <pre>
     A    C    G    T    
A  .990 .002 .006 .002
C  .002 .990 .002 .006
G  .006 .002 .990 .002
T  .002 .006 .002 .990
            </pre>
            </td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">--paths</td><td><span class="popt">single</span>|<span class="popt">all</span></td>
            <td>This option determines how the program computes raw scores.
            With the <span class="popt">single</span> option, the program
            computes the Viterbi score, which is the log-odds score associated
            with the single most likely match between the sequence and the
            model. The <span class="popt">all</span> option yields the total
            log-odds score, which is the sum of the log-odds of all
            sequence-to-model matches.</td>
            <td>Viterbi scoring is used.</td>
          </tr>
          <tr>
            <td class="popt">--p-score</td><td><span class="popt">num</span></td>
            <td>The <code>--p-score</code> switch activates <i>p</i>-value
            score mode with the given threshold. In p-value score mode, motif
            match scores are converted to their p-values. They are then
            converted to bit scores as follows:
            <div style="margin-left: 2em">S = -log<sub>2</sub>(p/T)</div>
            where S is the bit score of the hit, p is the p-value of the
            log-odds score, and T is the p-value threshold. In this way, only
            hits more significant than the p-value threshold get positive
            scores. The p-value threshold, T, must be in the range 0&lt;T&le;1.</td>
            <td>log-odds score mode is used.</td>
          </tr>
        </tbody>
      </table>
    </div>

  </body>
</html>

