<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>Citing - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>
    <h2>Citing MEME Suite Programs</h2>
    <h3>To cite the full MEME Suite</h3>
    <ul>
      <li>
        <div class="citation">
          Timothy L. Bailey, Mikael Bod&eacute;n, Fabian A. Buske, Martin Frith, Charles E. Grant, Luca Clementi,
          Jingyuan Ren, Wilfred W. Li, William S. Noble,
          &quot;MEME SUITE: tools for motif discovery and searching&quot;,
          <i>Nucleic Acids Research</i>, 37:W202-W208, 2009.
          <a
          href="http://nar.oxfordjournals.org/content/37/suppl_2/W202.full">[full text]</a>
        </div>
      </li>
    </ul>
    <h3>To cite individual tools</h3>
    <ul>
      <li>
        <b>AMA</b>
        <div class="citation">
          Fabian A. Buske, Mikael Bod&eacute;n, Denis C. Bauer and Timothy L. Bailey,
          &quot;Assigning roles to DNA regulatory motifs using comparative genomics&quot;,
          <i>Bioinformatics</i>, 26(7):860-866, 2010.
          <a href="http://bioinformatics.oxfordjournals.org/cgi/content/full/26/7/860">[full text]</a>
        </div>
      </li>
      <li>
        <b>AME</b>
        <div class="citation">
          Robert C. McLeay, Timothy L. Bailey,
          &quot;Motif Enrichment Analysis: a unified framework and an
          evaluation on ChIP data&quot;,
          <i>BMC Bioinformatics</i>,
          11:165, 2010.
          <a href="http://www.biomedcentral.com/1471-2105/11/165">[full text]</a>
        </div>
      </li>
      <li>
        <b>CentriMo</b>
        <div class="cite">
          Timothy L. Bailey and Philip Machanick, 
          &quot;Inferring direct DNA binding from ChIP-seq&quot;, 
          <i>Nucleic Acids Research</i>, <b>40</b>:e128, 2012.
          <a href="http://nar.oxfordjournals.org/content/40/17/e128">[Abstract and Full Text]</a>
        </div>
      </li>
      <li>
        <b>DREME</b>
        <div class="citation">
          Timothy L. Bailey,
          &quot;DREME: Motif discovery in transcription factor ChIP-seq data&quot;,
          <i>Bioinformatics</i>, 27(12):1653-1659, 2011.
          <a href="http://bioinformatics.oxfordjournals.org/content/27/12/1653">[full text]</a>
        </div>
      </li>
      <li>
        <b>FIMO</b>
        <div class="citation">
          Charles E. Grant, Timothy L. Bailey, and William Stafford Noble,
          &quot;FIMO: Scanning for occurrences of a given motif&quot;,
          <i>Bioinformatics</i> 27(7):1017&ndash;1018, 2011.
          <a href="http://bioinformatics.oxfordjournals.org/content/27/7/1017">[full text]</a>
        </div>
      </li>
      <li>
        <b>GLAM2 and GLAM2SCAN</b>
        <div class="citation">
          Martin C. Frith, Neil F. W. Saunders, Bostjan Kobe, Timothy L. Bailey,
          &quot;Discovering sequence motifs with arbitrary
          insertions and deletions&quot;,
          <i>PLoS Computational Biology</i>, <b>4</b>(5):e1000071, 2008.
          <a href="http://www.ploscompbiol.org/article/info:doi/10.1371/journal.pcbi.1000071">[full text]</a>
        </div>
      </li>
      <li>
        <b>GOMO</b>
        <div class="citation">
          Fabian A. Buske, Mikael Bod&eacute;n, Denis C. Bauer and Timothy L. Bailey,
          &quot;Assigning roles to DNA regulatory motifs using comparative genomics&quot;,
          <i>Bioinformatics</i>, 26(7), 860-866, 2010.
          <a href="http://bioinformatics.oxfordjournals.org/cgi/content/full/26/7/860">[full text]</a>
        </div>
      </li>
      <li>
        <b>MAST</b>
        <div class="citation">
          Timothy L. Bailey and Michael Gribskov,
          &quot;Combining evidence using p-values: application to sequence homology searches&quot;,
          <i>Bioinformatics</i>, 14(1):48-54, 1998.
          <a href="http://bioinformatics.oxfordjournals.org/content/14/1/48">[pdf]</a>
        </div>
      </li>
      <li>
        <b>MEME</b>
        <div class="citation">
          Timothy L. Bailey and Charles Elkan,
          &quot;Fitting a mixture model by expectation maximization to
          discover motifs in biopolymers&quot;,
          <em>Proceedings of the Second International Conference on Intelligent Systems
          for Molecular Biology</em>, pp. 28-36, AAAI Press, Menlo Park, California, 1994. 
          <a href="http://www.sdsc.edu/~tbailey/papers/ismb94.ps">[postscript]</a> 
          <a href="http://www.sdsc.edu/~tbailey/papers/ismb94.pdf">[pdf]</a>
        </div>
      </li>
      <li>
        <b>MEME-ChIP</b>
        <div class="citation">
          Philip Machanick and Timothy L. Bailey,
          &quot;MEME-ChIP: motif analysis of large DNA datasets&quot;,
          <i>Bioinformatics</i> 27(12):1696-1697, 2011.
        <a href="http://bioinformatics.oxfordjournals.org/content/27/12/1696.full">[full text]</a>
        </div>
      </li>
      <li>
        <b>PSPs</b>
        <div class="citation">
          Timothy L. Bailey, Mikael Bod&eacute;n, Tom Whitington, and Philip Machanick,
          &quot;The value of position-specific priors in motif
          discovery using MEME&quot;,
          <i>BMC Bioinformatics</i>, 11(1):179, 2010.
          <a href="http://www.biomedcentral.com/1471-2105/11/179">[full text]</a>
        </div>
      </li>
      <li>
        <b>MCAST</b>
        <div class="citation">
          Timothy Bailey and William Stafford Noble,
          &quot;Searching for statistically significant regulatory modules&quot;,
          <i>Bioinformatics (Proceedings of the European Conference on Computational Biology)</i>,
          19(Suppl. 2):ii16-ii25, 2003.
          <a href="http://bioinformatics.oxfordjournals.org/cgi/content/abstract/19/suppl_2/ii16">[full text]</a>
        </div>
      </li>
      <li>
        <b>SpaMo</b>
        <div class="citation">
          Tom Whitington, Martin C. Frith, James Johnson, and Timothy L. Bailey
          &quot;Inferring transcription factor complexes from ChIP-seq data&quot;,
          <i>Nucl. Acids Res.</i> 39(15):e98, 2011.
          <a href="http://nar.oxfordjournals.org/content/39/15/e98">[full text]</a>
        </div>
      </li>
      <li>
        <b>TOMTOM</b>
        <div class="citation">
          Shobhit Gupta, JA Stamatoyannopolous,
          Timothy Bailey and William Stafford Noble,
          &quot;Quantifying similarity between motifs&quot;,
          <i>Genome Biology</i>, 8(2):R24, 2007.
          <a href="http://genomebiology.com/2007/8/2/R24">[full text]</a>
        </div>
      </li>
    </ul>
    <h3>Related papers</h3>
    <h4>GOMO related</h4>
    <ul>
      <li class="citation">
        Mikael Bod&eacute;n and Timothy L. Bailey,
        &quot;Associating transcription factor binding site motifs with target Go terms and target genes&quot;,
        <em>Nucl. Acids Res</em>,
        36, 4108-4117, 2008.
        <a href="http://nar.oxfordjournals.org/cgi/content/abstract/gkn374v1">[full text]</a>
      </li>
    </ul>
    <h4>MAST related</h4>
    <ul>
      <li class="citation">
        Timothy L. Bailey and Michael Gribskov,
        &quot;Score distributions for simultaneous matching to multiple motifs&quot;
        <em>Journal of Computational Biology</em>,
        Vol. 4, pp. 45-59, 1997.
        <a href="http://www.sdsc.edu/~tbailey/papers/sumofgum.pdf">[pdf]</a>
      </li>
      <li class="citation">
        Timothy L. Bailey and Michael Gribskov, 
        &quot;Methods and statistics for combining motif match scores&quot;
        <em>Journal of Computational Biology</em>,
        Vol. 5, pp. 211-221, 1998.
        <a href="http://www.sdsc.edu/~tbailey/papers/stats.pdf">[pdf]</a>
      </li>
    </ul>
    <h4>MEME related</h4>
    <ul>
      <li class="citation">
        Timothy L. Bailey and Charles Elkan,
        &quot;Fitting a mixture model by expectation maximization to
        discover motifs in biopolymers&quot;,
        <em>Proceedings of the Second International Conference on Intelligent Systems
        for Molecular Biology</em>,
        pp. 28-36, AAAI Press, Menlo Park, California, 1994.
        <a href="http://www.sdsc.edu/~tbailey/papers/ismb94.ps">[postscript]</a>
        <a href="http://www.sdsc.edu/~tbailey/papers/ismb94.pdf">[pdf]</a>
      </li>
      <li class="citation">
        Timothy L. Bailey and Charles Elkan,
        &quot;Fitting a mixture model by expectation maximization to
        discover motifs in biopolymers&quot;,
        UCSD Technical Report, CS94-351, March, 1994, University of California at San Diego.
        <a href="http://www.sdsc.edu/~tbailey/papers/memeplus.tech.pdf">[pdf]</a>
      </li>
      <li class="citation">
        Timothy L. Bailey,
        &quot;Discovering motifs in DNA and protein sequences: The
        approximate common substring problem&quot;,
        Ph.D. dissertation, University of California at San Diego, 1995.
        <a href="http://www.sdsc.edu/~tbailey/papers/thesis.pdf">[pdf]</a>
      </li>
      <li class="citation">
        Timothy L. Bailey and Charles Elkan,
        &quot;Unsupervised Learning of Multiple Motifs in Biopolymers using EM&quot;
        <em>Machine Learning</em>,
        21(1-2):51-80, October, 1995.
        <a href="http://www.sdsc.edu/~tbailey/papers/meme.ml.pdf">[pdf]</a>
      </li>
      <li class="citation">
        Timothy L. Bailey and Charles Elkan,
        &quot;The value of prior knowledge in discovering motifs with MEME&quot;,
        <em>Proceedings of the Third International Conference on Intelligent Systems for Molecular Biology</em>,
        pp. 21-29, AAAI Press, Menlo Park, California, 1995.
        <a href="http://www.sdsc.edu/~tbailey/papers/ismb95.pdf">[pdf]</a>
      </li>
      <li class="citation">
        Timothy L. Bailey and Michael Gribskov,
        &quot;The megaprior heuristic for discovering protein sequence patterns&quot;",
        <em>Proceedings of the Fourth International Conference on Intelligent Systems for Molecular Biology</em>,
        pp. 15-24 AAAI Press, Menlo Park, California, 1996.
        <a href="http://research.imb.uq.edu.au/t.bailey/papers/ismb96.pdf">[pdf]</a>
      </li>
      <li class="citation">
        William N. Grundy, Timothy L. Bailey and Charles P. Elkan,
        &quot;ParaMEME: A Parallel Implementation and a Web Interface
        for a DNA and Protein Motif Discovery Tool&quot;,
        <em>Computer Applications in the Biological Sciences (CABIOS)</em>,
        Vol. 12(4), pp. 303-310, 1996.
        <a href="http://bioinformatics.oxfordjournals.org/content/12/4/303">[pdf]</a>
      </li>
      <li class="citation">
        Timothy L. Bailey, Michael E. Baker and Charles P. Elkan,
        &quot;An artificial intelligence approach to motif discovery in protein sequences:
        application to steriod dehydrogenases&quot;,
        <em>Journal of steroid biochemistry and molecular biology</em>,
        Vol. 62, 1997.
      </li>
      <li class="citation">
        Martin Tompa, Nan Li, Timothy L. Bailey, George M. Church, Bart De Moor,
        Eleazar Eskin, Alexander V. Favorov, Martin C. Frith, Yutao Fu, W. James Kent,
        Vsevolod J. Makeev, Andrei A. Mironov,, William S. Noble,
        Giulio Pavesi, Graziano Pesole, Mireille Regnier, Nicolas Simonis, Saurabh Sinha, 
        Gert Thijs, Jacques van Helden, Mathias Vandenbogaert, Zhiping Weng, Christopher Workman,
        Chun Ye and Zhou Zhu.
        &quot;Assessing computational tools for the discovery of
        transcription factor binding sites.&quot;  
        <em>Nature Biotechnology</em>,
        23(1), 137-144, 2005.
        <a href="http://www.nature.com/nbt/journal/v23/n1/full/nbt1053.html">[full text]</a>
      </li>
      <li class="citation">
        Timothy L. Bailey, Nadya Williams, Chris Misleh, and Wilfred W. Li,
        &quot;MEME: discovering and analyzing DNA and protein sequence motifs&quot;
        <em>Nucleic Acids Research</em>,
        Vol. 34, pp. W369-W373, 2006.
        <a href="http://nar.oxfordjournals.org/cgi/content/full/34/suppl_2/W369">[html]</a>
        <a href="http://nar.oxfordjournals.org/cgi/reprint/34/suppl_2/W369.pdf">[pdf]</a>
      </li>
    </ul>
    <h4>TOMTOM related</h4>
    <ul>
      <li class="citation">
        Emi Tanaka, Timothy L. Bailey, Charles E. Grant, William S. Noble, and Uri Keich,
        &quot;Improved similarity scores for comparing motifs&quot;
        <em>Bioinformatics</em>
        27(12): 1603-1609, 2011.
        <a href="http://bioinformatics.oxfordjournals.org/content/27/12/1603.full">[full text]</a>
      </li>
    </ul>
    <h4>Other</h4>
    <ul>
      <li class="citation">
        William N. Grundy, Timothy L. Bailey, Charles P. Elkan and Michael E. Baker.
        &quot;Hidden Markov Model Analysis of Motifs in Steroid Dehydrogenases and their Homologs&quot;
        <em>Biochemical and Biophysical Research Communications</em>,
        Vol 231, pp. 760-766, 1997.
        <a href="http://www.sdsc.edu/~tbailey/papers/dehydrog-bbrc97.pdf">[pdf]</a>
      </li>
      <li class="citation">
        William N. Grundy, Timothy L. Bailey, Charles P. Elkan and Michael E. Baker.
        &quot;Meta-MEME: Motif-based Hidden Markov Models of Protein Families&quot;
        <i>Computer Applications in the Biological Sciences (CABIOS)</i>,
        Vol. 13(4), pp. 397-406, 1997.
        <a href="http://bioinformatics.oxfordjournals.org/content/13/4/397">[pdf]</a>
      </li>
      <li class="citation">
        Michael E. Baker, William N. Grundy, and Charles P. Elkan,
        &quot;Spinach CSP41, an
        mRNA-binding protein and ribonuclease, is homologous to
        nucleotide-sugar epimerases and hydroxysteroid dehydrogenases&quot;,
        <em>Biochemical and Biophysical Research Communications</em>
        248(2), 250-254, 1998.
        <a href="http://www.sciencedirect.com/science?_ob=MImg&_imagekey=B6WBK-45KN84C-HN-1&_cdi=6713&_user=582538&_pii=S0006291X98989513&_origin=&_coverDate=07%2F20%2F1998&_sk=997519997&view=c&wchp=dGLzVlb-zSkWb&md5=d50511e1fb08f4347bf6a29bf745f91b&ie=/sdarticle.pdf">[pdf]</a>
      </li>
      <li class="citation">
        Michael E. Baker, William N. Grundy, and Charles P. Elkan,
        &quot;A common ancestor for a
        subunit in the mitochondrial proton-translocating NADH:ubiquinone
        oxidoreductase (complex I) and short-chain
        dehydrogenases/reductases&quot;,
        <em>Cellular and Molecular Life Sciences</em>,
        Vol. 55(3),  450-455, 1999.
      </li>
      <li class="citation">
        John Hawkins, Charles Grant, William S. Noble, and Timothy L. Bailey,
        &quot;Assessing
        phylogenetic motif models for predicting transcription factor binding sites&quot;
        <em>Bioinformatics
        (Proceedings of the Intelligent Systems for Molecular Biology Conference)</em>,
        25(12), i339--347, 2009.
        <a href="http://bioinformatics.oxfordjournals.org/content/25/12/i339.short">[full text]</a>
      </li>
      <li class="citation">
        Gabriel Cuellar-Partida, Fabian A. Buske, Robert C. McLeay, Tom Whitington,
        William Stafford Noble, and Timothy L. Bailey,
        &quot;Epigenetic priors for identifying active transcription 
        factor binding sites&quot;,
        <em>Bioinformatics</em> 28(1): 56-62, 2012
        <a href="http://bioinformatics.oxfordjournals.org/content/28/1/56.full.pdf">
        [pdf]
        </a>
      </li>
    </ul>
    <h3>Google Scholar Page</h3>
    <div class="indent">
      <a href="http://scholar.google.com/citations?user=4PFFWg0AAAAJ">MEME Suite on Google Scholar</a>
    </div>
  </body>
</html>

