<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>GLAM2 - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
    <script src="js/shared-doc.js"></script>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>
    <div class="prog_logo">
      <img src="images/glam2_icon.png" alt="GLAM2 Logo">
      <h1>GLAM2</h1>
      <h2>Gapped Local Alignment of Motifs</h2>
    </div>
    
    <div class="usage">
      <h3>Usage:</h3>
      <code>glam2 [options] &lt;alphabet&gt; &lt;sequences&gt;</code>
    </div>

    <h3>Description</h3>
    <div class="indent">
      <p>GLAM2 is a program for finding motifs in sequences, typically
      amino-acid or nucleotide sequences. The main innovation of
      GLAM2 is that it allows insertions and deletions in motifs.</p>
      <p>GLAM2 is a child of <a href="http://zlab.bu.edu/glam/">GLAM</a> and a
      sibling of <a href="http://www.ncbi.nlm.nih.gov/CBBresearch/Spouge/html.ncbi/index/software.html#7">A-GLAM</a>.
      'GLAM' used to stand for Gapless Local Alignment of Multiple sequences,
      but it now stands for Gapped Local Alignment of Motifs.</p>
      <p>GLAM2 uses a <a
      href="http://en.wikipedia.org/wiki/Simulated_annealing">simulated
      annealing</a> algorithm, with a temperature parameter. At high
      temperatures, GLAM2 only slightly favours changes that increase
      the alignment's score, and at low temperatures it strongly favours
      such changes. Thus, at high temperatures the score will be
      optimized too slowly, but at low temperatures the algorithm will
      get frozen in a local optimum. The strategy, then, is to start
      with a high temperature and reduce it as slowly as possible.</p>

      <p>Refer to the <a href="glam2_tut.html">GLAM2 tutorial</a> if you need
      more information.</p>
    </div>

    <h3>Inputs</h3>
    <div class="indent">
      <h4>Alphabet</h4>
      <p>The Alphabet of the sequences. This can be 'p' for protein sequences,
      'n' for nucleotide sequences or the name of an
      <a href="glam2_alphabet.html">GLAM2 alphabet file</a>.</p>
      <h4>Sequences</h4>
      <p>A file containing <a href="fasta-format.html">FASTA formatted sequences</a>.</p>
    </div>

    <h3>Outputs</h3>
    <div class="indent">
      <p>Outputs <code>glam2.html</code> and <code>glam2.txt</code> to the
      selected directory.</p>
    </div>

    <h3>Warnings</h3>
    <div class="indent">
    <p>GLAM2 might occasionally issue this warning: 'accuracy loss due
    to numeric underflow'. If this happens, its ability to optimise the
    alignment may be harmed: see <a href="GLAM2_method.pdf">GLAM2
    methods (PDF)</a> for details. To fix this, try increasing
    <span class="popt">-u</span>, or possibly decreasing
    <span class="popt">-b</span>. Increasing
    <span class="popt">-u</span> somewhat should be harmless, as long as it
    stays well below 1.</p>
    </div>

    <h3>Options</h3>
    <div class="indent">
      <table class="options">
        <thead>
          <tr>
            <th>Option</th>
            <th>Parameter</th>
            <th>Description</th>
            <th>Default Behaviour</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <th colspan="4">Input/Output</th>
          </tr>
          <script>
            wropt("all-o", {"dir": "glam2_out", "glam2": true});
            wropt("all-oc", {"dir": "glam2_out", "glam2": true});
          </script>
        </tbody>
        <tbody>
          <tr>
            <th colspan="4">Alignment Options</th>
          </tr>
          <tr>
            <td class="popt">-2</td><td></td>
            <td>Search both strands.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-z</td><td><span class="pdat">minimum sites</span></td>
            <td>Specify the minimum number sequences that must participate in
            the alignment. If greater than the number of input sequences, then
            all the sequences must participate.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-a</td><td><span class="pdat">minimum aligned</span></td>
            <td>Specify the minimum number of key positions (aligned columns)
            in the alignment.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-b</td><td><span class="pdat">maximum aligned</span></td>
            <td>Specify the maximum number of key positions (aligned columns)
            in the alignment.</td>
            <td></td>
          </tr>
        </tbody>
        <tbody>
          <tr>
            <th colspan="4">Scoring Scheme</th>
          </tr>
          <tr>
            <td class="popt">-d</td><td><span class="pdat">pseudocount</span></td>
            <td>Specify a <a href="http://compbio.soe.ucsc.edu/dirichlets/">Dirichlet mixture</a>
            file, which describes residues' tendencies to align with one another.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-D</td><td><span class="pdat">pseudocount</span></td>
            <td>Specify the deletion pseudocount.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-E</td><td><span class="pdat">pseudocount</span></td>
            <td>Specify the 'no-deletion' pseudocount.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-I</td><td><span class="pdat">pseudocount</span></td>
            <td>Specify the insertion pseudocount.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-J</td><td><span class="pdat">pseudocount</span></td>
            <td>Specify the 'no-insertion' pseudocount.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-q</td><td><span class="pdat">weight</span></td>
            <td>Specify the weight for generic versus sequence-set-specific
            residue abundances. The residue abundances are estimated by counting
            the residue types in all the input sequences, and adding pseudocounts.
            The total number of pseudocounts is equal to the alphabet size
            mutiplied by the <span class="popt">-q</span> parameter. The
            allocation of pseudocounts among the residue types depends on the
            <a href="glam2_alphabet.html">alphabet</a>.</td>
            <td></td>
          </tr>
        </tbody>
        <tbody>
          <tr>
            <th colspan="4">Search Algorithm</th>
          </tr>
          <tr>
            <td class="popt">-r</td><td><span class="pdat">num</span></td>
            <td>Specify the number of alignment runs.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-n</td><td><span class="pdat">n</span></td>
            <td>Specify how many iterations should pass since the
            highest-scoring alignment seen so far before ending each alignment
            run.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-w</td><td><span class="pdat">aligned columns</span></td>
            <td>Specify the initial number of key positions (aligned columns)
            in the alignment. If less than the minimum
            (<span class="popt">-a</span>) or greater than the maximum
            (<span class="popt">-b</span>), it is increased to the minimum or
            reduced to the maximum.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-t</td><td><span class="pdat">temperature</span></td>
            <td>Specify the initial temperature.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-c</td><td><span class="pdat">cooling factor</span></td>
            <td>Specify the cooling factor per <span class="pdat">n</span>
            iterations. The temperature is multiplied by a constant factor
            after each iteration, such that after <span class="pdat">n</span>
            iterations, it has dropped by this amount. The
            <span class="popt">-n</span> option is used to set the
            <span class="pdat">n</span> value.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-u</td><td><span class="pdat">minimum temperature</span></td>
            <td>Specify the minimum temperature. The temperature never drops
            below this level, to avoid numerical problems.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-m</td><td><span class="pdat">probability</span></td>
            <td>Specify the rate of column sampling relative to site sampling.
            On each iteration, glam2 randomly decides to try either realigning
            a sequence or adjusting an aligned column (key position). This
            parameter sets the probabilities for this decision.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-x</td><td><span class="popt">0</span>|<span
                class="popt">1</span>|<span class="popt">2</span></td>
            <td>Specify the site sampling algorithm: 0=FAST, 1=SLOW, 2=FFT.
            See <a href="GLAM2_method.pdf">GLAM2 methods (PDF)</a> for details. In
            summary, the FAST algorithm deviates slightly from the strict
            definition of simulated annealing, but works well in practice. The SLOW
            algorithm implements strict simulated annealing, but is much slower,
            especially for longer sequences. The FFT algorithm also implements
            strict simulated annealing, and has intermediate speed, but carries
            greater risk of numerical roundoff error. In order to use the FFT
            algorithm, it is necessary to install the <a
            href="http://www.fftw.org/">FFTW library</a>, and re-compile glam2
            with 'make glam2fft'.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-s</td><td><span class="pdat">seed</span></td>
            <td>Specify the seed for pseudo-random number generation. Change
            this to avoid getting identical results each time the program is
            run.</td>
            <td></td>
          </tr>
        </tbody>
        <tbody>
          <tr>
            <th colspan="4">Cosmetic</th>
          </tr>
          <tr>
            <td class="popt">-h</td><td>&nbsp;</td>
            <td>Show all options with their default settings and exit.</td>
            <td>Run as normal.</td>
          </tr>
          <tr>
            <td class="popt">-p</td><td>&nbsp;</td>
            <td>Print information about the algorithm's state before each
            iteration. The following information is printed: the temperature,
            the number of key positions in the alignment, the number of
            sequences in the alignment, and the alignment's score. In
            addition, some information about each move is printed: for site
            sampling moves, which sequence is picked, and for column sampling
            moves, which column is picked, whether or not it is deleted, and
            whether the direction is left or right.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-Q</td><td>&nbsp;</td>
            <td>Run quietly, suppressing unnecessary messages.</td>
            <td></td>
          </tr>
        </tbody>
      </table>
    </div>
    <h3>About</h3>
    <div class="indent">
      <p>GLAM2 was developed by <a
        href="http://www.cbrc.jp/~martin/">Martin C Frith</a>, working at the <a
        href="http://www.cbrc.jp/index.eng.html">Computational Biology
      Research Center</a> in Tokyo, and <a
        href="http://research.imb.uq.edu.au/~tbailey/">Timothy L Bailey</a>,
      working at the <a href="http://www.imb.uq.edu.au/">Institute for
      Molecular Bioscience</a> in Brisbane. The source code and documentation are
    hereby released into the public domain. If you use GLAM2, please cite:</p>
    <div class="citation">  
      <cite>
        MC Frith, NFW Saunders, B Kobe, TL Bailey (2008)
        Discovering sequence motifs with arbitrary insertions and deletions,
        PLoS Computational Biology, <b>4</b>(5), e1000071, 2008.
        <a href="http://www.ploscompbiol.org/article/info:doi/10.1371/journal.pcbi.1000071">[full text]</a>
      </cite>
      </div>
    </div>

  </body>
</html>

