<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>glam2format - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>
    
    <h2>glam2format</h2>
    
    <div class="usage">
      <h3>Usage:</h3>
      <code>glam2format [fasta|msf] &lt;glam2 alignment&gt;</code>
    </div>

    <h3>Description</h3>
    <div class="indent">
      <p>This program reads in a motif found by glam2, and writes it in a
      standard alignment format (FASTA-with-gaps or MSF). This enables the
      alignment to be passed to third-party software, including graphical
      visualization tools such as <a href="http://msa.cgb.ki.se/">Kalignvu</a>,
      <a href="http://www.ch.embnet.org/software/BOX_form.html">Boxshade</a>,
      and <a href="http://weblogo.berkeley.edu/">WebLogo</a>. On the other
      hand, not all the motif information is preserved: in particular, the key
      positions are lost. Only the top motif in glam2 output is converted.</p>
    </div>

    <h3>Inputs</h3>
    <div class="indent">
      <h4>Output Type</h4>
      <p>Choose to output sequences in FASTA-with-gaps or MSF.</p>
      <h4>GLAM2 Alignment</h4>
      <p>A GLAM2 motif.</h4>
    </div>

    <h3>Outputs</h3>
    <div class="indent">
      <p>The sequences are written to standard output unless
      <span class="popt">-o</span> is specified to send the output to a file.</p>
    </div>


    <h3>Options</h3>
    <div class="indent">
      <table class="options">
        <thead>
          <tr>
            <th>Option</th>
            <th>Parameter</th>
            <th>Description</th>
            <th>Default Behaviour</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <th colspan="4">General Options</th>
          </tr>
          <tr>
            <td class="popt">-o</td><td><span class="pdat">file</span></td>
            <td>The sequences are written to <span class="pdat">file</span>.</td>
            <td>The sequences are written to standard output.</td>
          </tr>
          <tr>
            <td class="popt">-c</td><td></td>
            <td>Make a compact alignment.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-f</td><td><span class="pdat">file</span></td>
            <td>Sequence file for flanking sequences</td>
            <td></td>
          </tr>
        </tbody>
      </table>
    </div>

    <h3>Usage Examples</h3>
    <div class="indent">
    <ul>
      <li>Basic usage: <pre>
glam2format fasta my_motif.glam2
</pre><pre>
glam2format msf my_motif.glam2
</pre></li>

      <li>Use -o to write the output to a file rather than to the screen: <pre>
glam2format -o my_motif.fa fasta my_motif.glam2
</pre><pre>
glam2format -o my_motif.msf msf my_motif.glam2
</pre></li>

      <li>Use -c to make a compact alignment: <pre>
glam2format -c fasta my_motif.glam2
</pre><pre>
glam2format -c msf my_motif.glam2
</pre> By default, residues that are inserted between key positions are
written as unaligned with each other. This best reflects glam2's
intention, but it can make the alignment large and full of gaps. With
-c, inserted residues are written as arbitrarily aligned with each
other, just as they appear in the glam2 output.</li>

<li>Use -f to make a 'global' alignment by adding flanking
sequences from the original FASTA-format sequence file: <pre>
glam2format -f my_seqs.fa fasta my_motif.glam2
</pre><pre>
glam2format -f my_seqs.fa msf my_motif.glam2
</pre> The flanking sequences will be written as either unaligned with each
other or arbitrarily aligned, depending on the -c option. The sequences
should have unique names and their order should be unchanged.</li>

</ul>

    </div>
  </body>
</html>

