<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>glam2scan - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>
    <div class="prog_logo">
      <img src="images/glam2scan_icon.png" alt="GLAM2Scan Logo">
      <h1>GLAM2Scan</h1>
      <h2>Scanning with Gapped Motifs</h2>
    </div>

    <div class="usage">
      <h3>Usage:</h3>
      <code>glam2scan [options] &lt;alphabet&gt; &lt;glam2 motif&gt; &lt;sequences&gt;</code>
    </div>

    <h3>Description</h3>
    <div class="indent">
      <p>GLAM2SCAN finds matches, in a sequence database, to a motif
      discovered by GLAM2. Each match receives a score, indicating how well it
      fits the motif.</p>
    </div>

    <h3>Input</h3>
    <div class="indent">
      <h4>Alphabet</h4>
      <p>The alphabet of the motif and sequences. This can be 'p' for protein
      sequences, 'n' for nucleotide sequences or the name of an
      <a href="glam2_alphabet.html">GLAM2 alphabet file</a>.</p>
      <h4>GLAM2 Motif</h4>
      <p>A file containing a GLAM2 motif. If the file contains multiple motifs
      then GLAM2SCAN only considers the top one.</p>
      <h4>Sequences</h4>
      <p>A file containing <a href="fasta-format.html">FASTA formatted sequences</a>.</p>
    </div>

    <h3>Output</h3>
    <div class="indent">
      <p>The output begins with some general information:</p>

      <pre>
GLAM2scan
Version 9999

glam2scan p prot_motif.glam2 lotsa_prots.fa
      </pre>

      <p>This is followed by motif matches, sorted in order of score. A
      motif match looks like this:</p>

      <pre>
                 **.****
SOS1_HUMAN   780 HPIE.IA 785 + 8.70
      </pre>

      <p>The name of the sequence with the match appears on the left; the
      start and end coordinates of the match appear on either side of the
      matching sequence; the match score appears on the right. The plus sign
      indicates the strand of the match (only meaningful when considering both
      strands of nucleotide sequences with the -2 option). The stars indicate
      the <em>key positions</em> of the motif: the alignment of the match to
      the key positions is shown.</p>
    </div>
    <h3>Options</h3>
    <div class="indent">
      <table class="options">
        <thead>
          <tr>
            <th>Option</th>
            <th>Parameter</th>
            <th>Description</th>
            <th>Default Behaviour</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <th colspan="4">Basic Options</th>
          </tr>
          <tr>
            <td class="popt">-o</td><td><span class="pdat">file</span></td>
            <td>Write the output to <span class="pdat">file</span>.</td>
            <td>Write the output to standard output.</td>
          </tr>
          <tr>
            <td class="popt">-n</td><td><span class="pdat">n</span></td>
            <td>Report <span class="pdat">n</span> matches. If scores are
            tied, they are sorted in alphabetical order of sequence name. If
            sequence names are also identical, the order is arbitrary.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-2</td><td></td>
            <td>Search both strands of nucleotide sequences.</td>
            <td></td>
          </tr>
        </tbody>
        <tbody>
          <tr>
            <th colspan="4">Advanced Options</th>
          </tr>
          <tr>
            <td colspan="4">The remaining options are somewhat specialized.
            For typical usage, it is reasonable to set them to exactly the
            same values as were used with glam2 to discover the motif.</td>
          </tr>
          <tr>
            <td class="popt">-D</td><td><span class="pdat">pseudocount</span></td>
            <td>Specify the deletion pseudocount.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-E</td><td><span class="pdat">pseudocount</span></td>
            <td>Specify the 'no-deletion' pseudocount.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-I</td><td><span class="pdat">pseudocount</span></td>
            <td>Specify the insertion pseudocount.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-J</td><td><span class="pdat">pseudocount</span></td>
            <td>Specify the 'no-insertion' pseudocount.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-d</td><td><span class="pdat">file</span></td>
            <td>Specify a <a href="http://compbio.soe.ucsc.edu/dirichlets/">Dirichlet mixture</a>
            file.</td>
            <td></td>
          </tr>
        </tbody>
      </table>
    </div>


    <h2>Motif format</h2>
      <div class="indent">
      <p>Some users may wish to make 'fake' glam2 motifs for input to
      glam2scan, for instance based on motifs found by other tools. Most of
      the glam2 output is ignored by glam2scan, and a minimal motif file looks
      like this:</p>

      <pre>
                **..****
seq1         10 HP..D.IG
seq2          5 HPGADLIG
seq3          7 HP..ELIG
seq4          5 HP..ELLA
      </pre>

      <p>The sequence names and coordinates are ignored, but some
      placeholder characters should be present. The stars indicating key
      positions are necessary, and the first and last columns must be starred.</p>
    </div>
<h3>Citing</h3>
<div class=indent>
<p>If you use this program in your research, please cite:
<div class=citation>
MC Frith, NFW Saunders, B Kobe, TL Bailey, &quot;Discovering sequence motifs with arbitrary insertions and deletions&quot;, PLoS Computational Biology, <b>4</b>(5):e1000071, 2008.
<a href="http://www.ploscompbiol.org/article/info:doi/10.1371/journal.pcbi.1000071">[full text]</a> </div>
</p>
</div

  </body>
</html>

