<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>psp-gen - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>
    <h2>psp-gen</h2>
    
    <div class="usage">
      <h3>Usage:</h3>
      <code>psp-gen -pos &lt;primary sequences&gt; -neg &lt;control sequences&gt; [options]</code>
    </div>

    <h3>Description</h3>
    <div class="indent">
      <p>psp-gen is used to allow MEME to perform discriminative motif
      discovery&mdash;to find motifs overrepresented in one set of sequences
      compared to in another set. It takes two files as input&mdash;the sequence
      file to be input to MEME, (the "primary" file) and a "control"
      sequence file of sequences believed not to contain the same motifs as
      in the "primary" file. psp-gen creates a file for use by MEME that
      encapsulates information about probable discriminative motifs. psp-gen
      records its chosen motif width in the file, and MEME is able to adjust
      the data when it tries different motif widths.</p>
    </div>

    <h3>Input</h3>
    <div class="indent">
      <h4>Primary Sequences File</h4>
      <div class="indent">
        <p>A file containing <a href="fasta-format.html">FASTA formatted sequences</a>
        which are to be used as the primary set in PSP calculation.</p>
      </div>

      <h4>Control Sequence File</h4>
      <div class="indent">
        <p>A file containing <a href="fasta-format.html">FASTA formatted sequences</a>
        which are to be used as the control set in PSP calculation.</p>
      </div>
    </div>

    <h3>Output</h3>
    <div class="indent">
      <p>A FASTA-like <a href="psp-format.html">PSP format</a> is written to
      standard output, it contains a prior for every position of every
      sequence in the primary set.</p>
    </div>

    <h3 id="options">Options</h3>
    <div class="indent">
      <table class="options">
        <thead>
          <tr>
            <th>Option</th>
            <th>Parameter</th>
            <th>Description</th>
            <th>Default Behaviour</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <th colspan="4">General Options</th>
          </tr>
          <tr>
            <td class="popt">-h</td><td></td>
            <td>Print a short usage message and exit.</td>
            <td>Run as normal.</td>
          </tr>
          <tr>
            <td class="popt">-minw</td><td><span class="pdat">min w</span></td>
            <td>The minimum width to use with selecting the "best" width for
              PSPs.</td>
            <td>The minimum width is set to 4.</td>
          </tr>
          <tr>
            <td class="popt">-maxw</td><td><span class="pdat">max w</span></td>
            <td>The maximum width to use with selecting the "best" width for
            PSPs.</td>
            <td>The maximum width is set to 20.</td>
          </tr>
          <tr>
            <td class="popt">-alpha</td><td><span 
                class="popt">dna</span>|&#8203;<span 
                class="popt">prot</td>
              <td>Specify the alphabet of the input sequences as DNA
                (<span class="popt">dna</span>) or protein 
                (<span class="popt">prot</span>).</td>
            <td>The DNA alphabet is assumed.</td>
          </tr>
          <tr>
            <td class="popt">-triples</td><td></td>
            <td>Use spaced triples instead of whole-word matches. Recommended for protein.</td>
            <td>Whole-word matches are used.</td>
          </tr>
          <tr>
            <td class="popt">-fixedstart</td><td></td>
            <td>When using the <span class="popt">-triples</span> option, only
            consider triples starting at the start of the site.</td>
            <td>Allow triples to start anywhere in a width w site.</td>
          </tr>
          <tr>
            <td class="popt">-equiv</td><td><span class="pdat">Letter Groups</span></td>
            <td>Any sequence of letter that appears together should be matched
              as equal. Separate letter groups using "-", e.g.
              <span class="popt">-equiv "IVL-HKR"</span> means treat all
              occurrences of I, V or L as the same, and all occurrences of H, K or R as
              the same.</td>
            <td>No letter groups are used.</td>
          </tr>
          <tr>
            <td class="popt">-revcomp</td><td></td>
            <td>Both strands are considered when calculating PSPs for DNA.</td>
            <td>Only the given strand is considered.</td>
          </tr>
          <tr>
            <td class="popt">-scalemin</td><td><span class="pdat">min score</span></td>
            <td>The lowest score value after scaling.</td>
            <td>The lowest score is set to 0.1, unless the
            <span class="popt">-scalemax</span> option is set in which case
            the lowest score is <span class="cmd"><span class="pdat">max score</span> - 1</span>.</td>
          </tr>
          <tr>
            <td class="popt">-scalemax</td><td><span class="pdat">max score</span></td>
            <td>The highest score value after scaling.</td>
            <td>The highest score is set to 0.9, unless the
              <span class="popt">-scalemin</span> option is set in which case
              the highest score is <span class="cmd"><span class="pdat">min score</span> + 1</span>.</td>
            </td>
          </tr>
          <tr>
            <td class="popt">-maxrange</td><td></td>
            <td>Choose the width with the biggest difference between minimum
              and maximum scores (before scaling).</td>
            <td>Choose the width with the biggest maximum score (before scaling).</td>
          </tr>
          <tr>
            <td class="popt">-raw</td><td></td>
            <td>Output scores instead of priors.</td>
            <td>The program output PSPs.</td>
          </tr>
          <tr>
            <td class="popt">-reportscores</td><td></td>
            <td>Send to standard error primary and control file names, min
              and max scores and min and max widths, tab-separated.</td>
            <td>Do not report scores to standard error.</td>
          </tr>
          <tr>
            <td class="popt">-verbose</td><td></td>
            <td>Send detailed stats to standard error, reporting frequency of each score.</td>
            <td>No extra information is reported.</td>
          </tr>
        </tbody>
      </table>
    </div>


    <h3>Examples</h3>
    <div class="indent">
      <p>In each example, the sequence file is the same as the postive set used
      for <code>psp-gen</code>. While there is no check that you have done
      this because MEME runs as a separate program (other than that the
      sequence names need to match, as do the number of sites), using a
      primary set different to the sequence set MEME searches is unlikely
      to be useful.</p>

      <p>Generate a discriminative prior for a DNA sequence set
      (<code>pos-dna.fasta</code>) likely to contain TF binding sites, given
      another DNA sequence set unlikely to contain TF binding sites
      (<code>neg-dna.fasta</code>): <br />
      <code>psp-gen -pos pos-dna.fasta -neg neg-dna.fasta &gt; dna.psp</code>
      </p>

      <p>Use these position-specific priors in MEME, searching both strands:<br />
      <code>meme pos-dna.fasta -psp dna.psp -oc meme_out_dna
      -revcomp</code><br />
      </p>


      <p>Generate a discriminative prior for a protein sequence set
      (<code>pos-prot.fasta</code>) associated with a specific function, given
      another protein sequence set unlikely to relate to that function
      (<code>neg-prot.fasta</code>): <br />
      <code>psp-gen -pos pos-prot.fasta -neg neg-prot.fasta -alpha prot
      -triples -maxrange &gt; prot.psp</code>
      </p>

      <p>Use these position-specific priors in MEME:<br />
      <code>meme pos-prot.fasta -psp prot.psp -oc meme_out_prot</code>
      </p>

      <p>View usage information:<br />
      <code>psp-gen -h</code>
      </p>
    </div>

    <h3 id="detail">Detailed description</h3>
    <div class="indent">

      <p>The psp-gen program generates a position-specific
      prior (PSP) in a data file, which can be used as an additional input
      to MEME to bias the search to sites more likely to
      contain a motif (or motifs) of interest. Without using a PSP, MEME
      assigns the same prior probability to all sites. A PSP is generated
      by <code>psp-gen</code> using two sequence sets, each
      in <a href="fasta-format.html">FASTA format</a>:
      <ul>
        <li>the <i>primary set</i> (or <i>X</i>) contains the same
          sequences as you intend to search for a motif using MEME</li>
        <li>the <i>control set</i> (or <i>Y</i>) is a contrasting set of
          sequences that is unlikely to contain the motif or motifs of
          interest, but is otherwise similar to the primary set.</li>
      </ul>
      </p>

      <p>The basic principle of a discriminative prior is to start from the
      question:</p>
      <div class="indent">
        <p>What fraction of <i>all</i> instances of a <i>w</i>-wide
        subsequence (or <i>w</i>-mer) across both sequence sets occur <i>in the
        primary set</i>?
        </p>
      </div>

      <p>The intuition is that any <i>w</i>-mer that is common
      in the primary set but not the control set could form part of motif
      of interest, based on the way the two sets are chosen.
      </p>

      <p>PSP calculation starts from the following equation based on the
      <i>D</i>-prior described in the Narlikar <i>et al</i>. <i>RECOMB</i>
      2007 paper
      "<a href="http://www.springerlink.com/content/av526j7u275n1508/">Nucleosome
      Occupancy Information Improves <i>de novo</i> Motif
      Discovery</a>". We count the number of instances 
      <i>X</i><sub><i>i</i>,<i>j</i></sub>(<i>w</i>) of the
      subsequence <i>w</i> wide starting at position <i>j</i> in sequence
      <i>X<sub>i</sub></i> of the primary set, and similarly
      count <i>Y</i><sub><i>i</i>,<i>j</i></sub>(<i>w</i>) in the control 
      set, then calculate for each site
<!-- begin MathToWeb -->
<!-- (your LaTeX) \begin{equation*}
D_{i,j} = \frac {X_{i,j}(w)} {X_{i,j}(w) + Y_{i,j}(w)}.
\end{equation*} -->
<math xmlns="http://www.w3.org/1998/Math/MathML">
<mtable class="m-equation-starred" style="width: 98%; margin-top: 1.0em; margin-right: 1.0em; margin-bottom: 2.0em; margin-left: 2.0em" displaystyle="true">
	<mtr>
		<mtd columnalign="left">
			<msub>
				<mi>D</mi>
				<mrow>
					<mi>i</mi>
					<mo>,</mo>
					<mi>j</mi>
				</mrow>
			</msub>
			<mo>=</mo>
			<mfrac linethickness="1">
				<mrow>
					<msub>
						<mi>X</mi>
						<mrow>
							<mi>i</mi>
							<mo>,</mo>
							<mi>j</mi>
						</mrow>
					</msub>
					<mrow>
						<mo maxsize="1.00em">(</mo>
						<mi>w</mi>
						<mo maxsize="1.00em">)</mo>
					</mrow>
				</mrow>
				<mrow>
					<msub>
						<mi>X</mi>
						<mrow>
							<mi>i</mi>
							<mo>,</mo>
							<mi>j</mi>
						</mrow>
					</msub>
					<mrow>
						<mo maxsize="1.00em">(</mo>
						<mi>w</mi>
						<mo maxsize="1.00em">)</mo>
					</mrow>
					<mo>+</mo>
					<msub>
						<mi>Y</mi>
						<mrow>
							<mi>i</mi>
							<mo>,</mo>
							<mi>j</mi>
						</mrow>
					</msub>
					<mrow>
						<mo maxsize="1.00em">(</mo>
						<mi>w</mi>
						<mo maxsize="1.00em">)</mo>
					</mrow>
				</mrow>
			</mfrac>
			<mo>.</mo>
		</mtd>
	</mtr>
</mtable>
</math>
<!-- end MathToWeb -->
      As pseudocounts, we add 1 to the enumerator, and 1 +
      |<i>X</i>|/|<i>Y</i>| to the denominator. The purpose of the
      pseudocounts is to avoid giving a high score to infrequent
      subsequences that are not (or rarely) found in the control set.  Once
      we have scored all sites for a given value of <code>w</code>, we scale
      scores to a range that can be <a href="#scaling">set on the command
      line</a>, with the default range <code>[0.1,0.9]</code>.</p>

      <p>We repeat scoring for the whole range of motif widths under
      consideration, and choose a "best" width using one of two methods:
      either we choose the value of <code>w</code> that maximizes the score
      over all sequences, or the value of <code>w</code> that maximizes the
      difference between the maximum and minimum score over all
      sequences. We also allow an option of using <i>spaced triples</i>
      instead of whole words to match
      subsequences. See <a href="#variations">Variations</a> below for more
      details on triples and choosing the "best" width.</p>

      <p>Once we have scored every available site and chosen a specific value
      of <code>w</code>, we convert the scores to probabilities, based on the
      proportionalities
<!-- begin MathToWeb -->
<!-- (your LaTeX) \begin{equation*}
\mbox{Pr}(Z_{i} = 0) \propto 1
\mbox{and}
\mbox{Pr}(Z_i = j) \propto \frac{D_{i,j}}{1-D_{i,j}},
\end{equation*} -->
<math xmlns="http://www.w3.org/1998/Math/MathML">
<mtable class="m-equation" style="width: 98%; margin-top: 1.0em; margin-right: 1.0em; margin-bottom: 2.0em; margin-left: 2.0em" displaystyle="true">
	<mtr>
		<mtd columnalign="left">
			<mtext>Pr</mtext>
			<mrow>
				<mo maxsize="1.00em">(</mo>
				<msub>
					<mi>Z</mi>
					<mi>i</mi>
				</msub>
				<mo>=</mo>
				<mn>0</mn>
				<mo maxsize="1.00em">)</mo>
			</mrow>
			<mo>&prop;</mo>
			<mn>1</mn>
			<mtext>&nbsp; and &nbsp; </mtext>
			<mtext>Pr</mtext>
			<mrow>
				<mo maxsize="1.00em">(</mo>
				<msub>
					<mi>Z</mi>
					<mi>i</mi>
				</msub>
				<mo>=</mo>
				<mi>j</mi>
				<mo maxsize="1.00em">)</mo>
			</mrow>
			<mo>&prop;</mo>
			<mfrac linethickness="1">
				<mrow>
					<msub>
						<mi>D</mi>
						<mrow>
							<mi>i</mi>
							<mo>,</mo>
							<mi>j</mi>
						</mrow>
					</msub>
				</mrow>
				<mrow>
					<mn>1</mn>
					<mo>-</mo>
					<msub>
						<mi>D</mi>
						<mrow>
							<mi>i</mi>
							<mo>,</mo>
							<mi>j</mi>
						</mrow>
					</msub>
				</mrow>
			</mfrac>
			<mo>,</mo>
		</mtd>
	</mtr>
</mtable>
</math>
<!-- end MathToWeb -->
      where <i>Z<sub>i</sub></i> = <i>j</i> means
      that the site for a motif in sequence <i>X<sub>i</sub></i> is found at
      location <i>j</i> (with <i>j</i> = 0
      signifying there is no site in the given sequence),
      and <i>D</i><sub><i>i</i>,<i>j</i></sub> is the score for
      sequence <i>X<sub>i</sub></i> at the site starting at
      location <i>j</i>, numbered from 1.
      We convert these proportionalities to a PSP by normalizing them to add
      up to 1 over each sequence.</p>
    </div>

    <h3 id="vsMEME">Running with MEME</h3>
    <div class="indent">
      <p>Since MEME runs as a separate program, there is no way to check that
      the primary set is the same as the sequence set given to MEME. There
      are however checks that every name in the PSP file matches a name in
      the MEME sequence set, and that the number of sites in the PSP file
      for a particular sequence name matches the number of sites in the
      sequence file seen by MEME. While it does not make sense to use a
      completely different sequence set for PSP generation and for running
      MEME (with the resulting PSP file), MEME can use a PSP file that does
      not provide probabilities for every sequence. Any sequences not in the
      PSP file are given a uniform prior
      probability. <a href="meme.html#MEME-PSP">MEME documentation</a>
      includes more detail on how PSPs are used.</p>
    </div>

    <h3 id="variations">Variations</h3>
    <div class="indent">
      <p>For each variation, "default" means that you do not need to specify a
      command-line switch to turn on that setting.</p>

      <ul>
        <li><i>sequence alphabet</i>: choose <span class="popt">-alpha DNA</span>
        (default) or <span class="popt">-alpha prot</span> (protein). For DNA, you also
        have the option to score matches on both strands:
           <ul>
             <li><code>-revcomp</code> (the default is to score only the given
             strand). This setting applies to both the primary and control 
             sequences.
             </li>
           </ul>
        </li>
        <li><i>choosing the "best" width</i>: specifying <code>-maxw</code>
        and <code>-minw</code> provides limits to <code>psp-gen</code> but
        the algorithm chooses one width to generate priors after scoring the
        range of allowed widths, using one of two variations:
        </li>
        <ul>
           <li><i>maximum score</i> (default): choose the value of <i>w</i> that
           maximizes the score over all sequences.
          </li>
          <li><i>maximum range of values</i> (<code>-maxrange</code>):
          choose the value of <i>w</i> that maximizes the difference between
          the highest and lowest score over all sequences. This variant is
          useful in data sets where a high number of sites achieve similar
          high scores (more likely to happen with protein sequences).
          </li>
        </ul>
        <li><i>type of word match</i>:
        <ul>
          <li><i>whole-word match</i> (default)
          </li>
          <li><i>spaced triples</i> (<code>-triples</code>): for a case
           where exact-word matches are unlikely to succeed, spaced triples
           are an alternative. Each <i>w</i>-wide word is broken down into
           triples containing an initial, middle and final letter, and all
           other letters are treated as don't cares (i.e., they match
           anything). In spaced triples mode, <code>psp-gen</code> scores a
           site on the basis of the spaced triple fitting the <i>w</i>-wide
           word at that site that has the highest score.  For example, the
           subsequence<br />
           <code>MTFEKI</code> contains the following
           triples:<br />
           <code>MT...I</code><br />
           <code>M.F..I</code><br />
           <code>M..E.I</code><br />
           <code>M...KI</code><br />
           where "." matches anything when counting matches.
           <ul>
             <li>An option that only applies to triples
               is <code>-fixedstart</code>, which only allows triples for the
               current site under consideration that start at the beginning of the
               site. This option reduces execution time at the expense of
               considering fewer possibilities for scoring a site.
             </li>
             <li>default: consider triples of all widths that fit the current site.
             </li>
           </ul>
          </li>
        </ul>
        </li>
        <li id="scaling"><i>scaling variations</i> (<code>-scalemin</code> and
        <code>-scalemax</code>): the scoring formula creates values in the
        range <code>[0,1]</code>, but the conversion to PSPs from scores
        requires values strictly less than 1. Choosing an option other than
        scaling between the default <code>[0.1,0.9]</code> reduces (if you
        narrow the range) or increases (if you widen the range) sensitivity
        to small differences in scores. If you only set one
        of <code>scalemin</code> or <code>scalemax</code>, the other is
        calculated as described in <a href="#options">Options</a>.
        </li>
        <li><i>ambiguous characters</i>: you have two levels of control of
        ambiguous characters:
        <ul>
          <li><i>ambiguous characters in the sequence data</i>:
          The effect of encountering any of these letters is to score a word
          or triple containing them as if it only occurs once, to minimize its
          score.
          </li>
          <li><i>treat specific letters as equivalent</i>: use
          the <code>-equiv</code> option to treat specific groups of letters
          as if they were all the same (default: all letters are treated as
          unique). For more detail, see <a href="#options">Options</a>.
          </li>
        </ul>
        </li>
        <li><i>reporting and debugging</i>: you have various options for
        understanding better how PSPs are calculated.
        <ul>
          <li><i>see raw scores</i>: <code>-raw</code> stops the last stage
          of computation that converts the scores to probabilities; the
          scores sent to standard out are however scaled, using either the
          defaults or selected values of <code>scalemin</code>
          and <code>scalemax</code> (default: convert scores to PSPs).
          </li>
          <li><i>report range of scores</i>:
          see <code>-reportscores</code> in <a href="#options">Options</a>
          (default: do not report).
          </li>
          <li><i>see detailed statistics</i>: <code>-verbose</code> causes
          detailed statistics on scores to be sent to standard error
          (default: no statistics reported).
          </li>
        </ul>
        </li>
      </ul>
    </div>

    <h3>Performance scalability</h3>
    <div class="indent">
      <p>When you use exact-word (<i>w</i>-mer) search, psp-gen
      run time scales linearly in each of |<i>X</i>|, |<i>Y</i>| and
      <code>maxw</code>-<code>minw</code>. Formally, time is O((|<i>X</i>| +
              |<i>Y</i>|)(<i>w<sub>max</sub></i> -
              <i>w<sub>min</sub></i>)).
      For triples, <code>psp-gen</code> scales linearly in each of
              |<i>X</i>| and |<i>Y</i>|, and quadratically
              in <i>w<sub>max</sub></i>, so run time is O((|<i>X</i>| +
              |<i>Y</i>|)<i>(w<sub>max</sub></i>)<sup>2</sup>).</p>
      <p>Extra memory in both cases scales linearly with |<i>X</i>| +
              |<i>Y</i>|.</p>
      <p>
      Since triples have a higher-order term in time complexity, we predict
              run time scalability in more detail than for exact word
              match. Adding in constants and a linear term, time complexity
              for triples is approximated by (where <i>W</i>
              = <i>w<sub>max</sub></i>-2, the number of variations in width
              used by the algorithm):
      </p>
      <table border="0">
      <tr><td><i>t</i></td>
        <td>=</td>
        <td>(<i>k</i><sub>1</sub>|<i>X</i>| +
          <i>k</i><sub>2</sub>|<i>Y</i>|)<i>W</i><sup>2</sup> +
          <i>k</i><sub>3</sub>(|<i>X</i>| + |<i>Y</i>|)<i>W</i>
        </td></tr>
      </table>
      <p>
      On a real machine (iMac, 3.06GHz Intel Core 2 Duo, 6MB L2 cache, 4GB
      RAM) the following is a good predictor of the upper bound on run time
      for likely values of |<i>X</i>|, |<i>Y</i>| and range of
      widths. Scale any results according to the speed of the actual machine
      on which you intend to run <code>psp-gen</code>. The formula reflects
      runs on randomly generated data; real data will likely result in
      faster runs because random data defeats some optimizations
      in <code>psp-gen</code>.
      </p>
      <table border="0">
      <tr><td><b><i>t</i></b></td><td><b>=</b></td><td><b>(4.37x10<sup>-6</sup>|<i>X</i>|+
            2.46x10<sup>-06</sup>|<i>Y</i>|)<i>W</i><sup>2</sup> +
            6.25x10<sup>-06</sup>(|<i>X</i>| + |<i>Y</i>|)<i>W</i> s</b></td></tr>
      </table>
      <p>
      For example, for |<i>X</i>| = |<i>Y</i>| = 60,000
      and  <i>w<sub>max</sub></i> = 50 (i.e., <i>W</i> = 48), estimated run
      time = 980s, the same in this case as measured run time. The following
      command line reproduces this run:
      </p>
      <pre>
      time psp-gen -neg pos60k.fasta -pos neg60k.fasta -alpha prot -minw 3 -maxw 50 -triples &gt; priors60kmaxw50.psp
      </pre>
    </div>

    <h3>Known problems</h3>
    <div class="indent">
      <ul>
        <li>
        Any letters in sequence represening ambiguous bases or amino acids
        result in a word (<i>w</i>-mer) containing that word (or triple in
        the case of protein) being counted whether in the primary or
        control sets as if it occurred once, to minimize its score.
        </li>
      </ul>
    </div>
  </body>
</html>

