<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>purge - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>
    <h2>purge</h2>
    
    <div class="usage">
      <h3>Usage:</h3>
      <code>purge &lt;sequence file&gt; &lt;score threshold&gt; [options]</code>
    </div>

    <h3>Description</h3>
    <div class="indent">
      <p>Purge removes redundant sequences from a FASTA file. This is
      recommended in order to prevent highly similar sequences distorting the
      search for motifs. Purge was written by Andy Neuwald and is described in
      more detail in Neuwald <i>et al.</i>, "Gibbs motif sampling: detection
      of bacterial outer membrane protein repeats", Protein Science,
      <b>4</b>:1618--1632, 1995.</p>

      <p>Purge works with either DNA or protein sequences and creates an
      output file such that no two sequences have a (gapless) local alignment
      score greater than a threshold specified by the user. The alignment
      score is based on the BLOSUM62 matrix for proteins, and on a +5/-1
      scoring scheme for DNA. Purge can also be used to mask tandem repeats.
      It uses the XNU program for this purpose.</p>
    </div>

    <h3>Input</h3>
    <div class="indent">
      <h4>Sequence File</h4>
      <div class="indent">
        <p>A file containing <a href="fasta-format.html">FASTA formatted sequences</a>.
        This file will be overwritten!</p>
      </div>

      <h4>Score Threshold</h4>
      <div class="indent">
        <p>The minimum local alignment score threshold for discarding similar sequences.</p>
      </div>
    </div>

    <h3>Output</h3>
    <div class="indent">
      <p>Purge overwrites the original <a href="fasta-format.html">FASTA sequence file</a>
      with the purged sequences.</p>
    </div>

    <h3>Options</h3>
    <div class="indent">
      <table class="options">
        <thead>
          <tr>
            <th>Option</th>
            <th>Parameter</th>
            <th>Description</th>
            <th>Default Behaviour</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <th colspan="4">General Options</th>
          </tr>
          <tr>
            <td class="popt">-n</td><td></td>
            <td>Sequences are treated as DNA.</td>
            <td>Sequences are treated as protein.</td>
          </tr>
          <tr>
            <td class="popt">-b</td><td></td>
            <td>Use BLAST heuristic method. This is not compatible with
              DNA sequences.</td>
            <td>Protein sequences use the BLAST heuristic method and DNA
              sequences use the exhaustive method.</td>
          </tr>
          <tr>
            <td class="popt">-e</td><td></td>
            <td>Use an exhaustive method.</td>
            <td>Protein sequences use the BLAST heuristic method and DNA
              sequences use the exhaustive method.</td>
          </tr>
          <tr>
            <td class="popt">-q</td><td></td>
            <td>Always keep the first sequence in the set.</td>
            <td>The first sequence may be removed.</td>
          </tr>
          <tr>
            <td class="popt">-x</td><td></td>
            <td>Use the XNU program to mask protein tandem repeats. This
              option is only compatible with protein sequences.</td>
            <td>Protein tandem repeats are not masked.</td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>
</html>

